### PR TITLE
Feat/pricer fixes

### PR DESCRIPTION
**Pull request by aparnakr**
# Task: Pricer fixes

## Enable anyone to set prices for all assets

In the setExpiryPriceInOracle function, if anyone is calling the function, checks that the expiryTimestamp is less than the expiry timestamp of the passed in roundId. Also checks that the expiryTimestamp is grater than the expiry timestamp of the previous roundId. This ensures that someone cannot pass in an old price or too new of a price. 

### Code

- [x] Unit test 100% coverage
- [x] Does your code follow the naming and code documentation guidelines?

### CODE REVIEW

File: contracts/pricers/ChainlinkPricer.sol [hunk 1]
```diff
@@ -52,26 +52,34 @@ contract ChainLinkPricer is OpynPricerInterface {
...
      */
-    function setExpiryPriceInOracle(uint256 _expiryTimestamp, uint80 _roundId) external onlyBot {
+    function setExpiryPriceInOracle(uint256 _expiryTimestamp, uint80 _roundId) external {
         (, int256 price, , uint256 roundTimestamp, ) = aggregator.getRoundData(_roundId);
 
         require(_expiryTimestamp <= roundTimestamp, "ChainLinkPricer: invalid roundId");
 
+        if (msg.sender != bot) {
+            bool isCorrectRoundId;
+            uint80 previousRoundId = uint80(uint256(_roundId).sub(1));
```
Comment (antoncoding) [inline contracts/pricers/ChainlinkPricer.sol:68]: I think we  can just use roundId - 1, not 2 extra casting.
If it overflows, it's a big roundId that Chainlink won't return anything, and there's no way in get back to a realistic roundId with the gas limit, from `uint80(-1)`

**Review by antoncoding — APPROVED** [review]

**Review by aleone — APPROVED** [review]
Generally lgtm, think we should change revert message to be distinct for each require. I also notice the build is failing re tests/coverage on github.

File: contracts/pricers/ChainlinkPricer.sol (hunk 1 above)
Comment (haythem96) [reply contracts/pricers/ChainlinkPricer.sol:68]: I agree, I also think we will not have `_roundId = 0` where `- 1` will underflow

File: contracts/pricers/ChainlinkPricer.sol [hunk 2]
```diff
@@ -52,26 +52,34 @@ contract ChainLinkPricer is OpynPricerInterface {
...
+
+            while (!isCorrectRoundId) {
+                (, , , uint256 previousRoundTimestamp, ) = aggregator.getRoundData(previousRoundId);
+
+                if (previousRoundTimestamp == 0) {
+                    previousRoundId = uint80(uint256(previousRoundId).sub(1));
+                } else if (previousRoundTimestamp > _expiryTimestamp) {
+                    revert("ChainLinkPricer: invalid roundId");
```
[the remark this answers is not in the document]
Comment (haythem96) [reply contracts/pricers/ChainlinkPricer.sol:76]: Agree, we should update messages.

File: contracts/pricers/ChainlinkPricer.sol [hunk 3]
```diff
@@ -52,26 +52,34 @@ contract ChainLinkPricer is OpynPricerInterface {
...
-
     /**
      * @notice set the expiry price in the oracle, can only be called by Bot address
      * @dev a roundId must be provided to confirm price validity, which is the first Chainlink price provided after the expiryTimestamp
      * @param _expiryTimestamp expiry to set a price for
      * @param _roundId the first roundId after expiryTimestamp
      */
-    function setExpiryPriceInOracle(uint256 _expiryTimestamp, uint80 _roundId) external onlyBot {
+    function setExpiryPriceInOracle(uint256 _expiryTimestamp, uint80 _roundId) external {
         (, int256 price, , uint256 roundTimestamp, ) = aggregator.getRoundData(_roundId);
```
[the remark this answers is not in the document]
Comment (haythem96) [reply contracts/pricers/ChainlinkPricer.sol:62]: I think we should add a require here for that

File: test/unit-tests/chainlinkPricer.test.ts [hunk 4]
```diff
@@ -128,15 +128,32 @@ contract('ChainlinkPricer', ([owner, bot, random]) => {
       assert.equal(p1.toString(), priceFromOracle[0].toString())
     })
 
-    it('should revert if sender is not bot address', async () => {
+    it('should revert if sender is not bot address and roundId is too early', async () => {
       const expiryTimestamp = (t1 + t2) / 2 // between t0 and t1
```
Comment (haythem96) [inline test/unit-tests/chainlinkPricer.test.ts:132]: Updatet the comment here

File: test/unit-tests/chainlinkPricer.test.ts [hunk 5]
```diff
@@ -128,15 +128,32 @@ contract('ChainlinkPricer', ([owner, bot, random]) => {
...
       await expectRevert(
-        pricer.setExpiryPriceInOracle(expiryTimestamp, roundId, { from: random }),
-        'ChainLinkPricer: unauthorized sender',
+        pricer.setExpiryPriceInOracle(expiryTimestamp, roundId, {from: random}),
+        'ChainLinkPricer: invalid roundId',
       )
     })
 
+    it('should revert if sender is not bot address and roundId is too late', async () => {
+      const expiryTimestamp = (t1 + t2) / 2 // between t0 and t1
```
Comment (haythem96) [inline test/unit-tests/chainlinkPricer.test.ts:141]: Update comment here also

File: test/unit-tests/chainlinkPricer.test.ts [hunk 6]
```diff
@@ -128,15 +128,32 @@ contract('ChainlinkPricer', ([owner, bot, random]) => {
...
+      const roundId = 3
+      await expectRevert(
+        pricer.setExpiryPriceInOracle(expiryTimestamp, roundId, {from: random}),
+        'ChainLinkPricer: invalid roundId',
+      )
+    })
+
+    it('anyone should be able to set prices', async () => {
+      const expiryTimestamp = (t1 + t2) / 2 // between t0 and t1
```
Comment (haythem96) [inline test/unit-tests/chainlinkPricer.test.ts:150]: Update comment here

**Review by haythem96 — APPROVED** [review]
Approved for now,  left few comments, need to fix build also.

**Comment by aparnakr** [comment]
> Approved for now, left few comments, need to fix build also.

build is broken on master. not sure why :( Maybe @antoncoding do you know?

File: scripts/README.md [hunk 7]
```diff
@@ -26,7 +26,7 @@ To deploy a new `ChainlinkPricer.sol`, it is recommended to use the `deployChain
 **Input**
 
 ```sh
-truffle exec scripts/deployChainlinkPricer.js --network mainnet --controller 0x7d78c401c69c56cb21f4bf80c53afd92be0BBBBB --pool 0xc02aaa39b223fe8d0a0e5c4f27ead9083c7AAaaa --weth 0x5f4eC3Df9cbd43714FE2740f5E3616155cAGAGAG --oracle 0xef196aA0e3Cb8EA6d5720557C3B611Eff6OOOOOO --gasPrice 50000000000
+truffle exec scripts/deployChainlinkPricer.js --network mainnet --bot 0xfacb407914655562d6619b0048a612B1795dF783 --asset 0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599 --aggregator 0xF4030086522a5bEEa4988F8cA5B36dbC97BeE88c --oracle 0x789cD7AB3742e23Ce0952F6Bc3Eb3A73A0E08833 --gasPrice 100000000000
```
Comment (haythem96) [inline scripts/README.md:29]: Is this command correct ?!

File: scripts/README.md [hunk 8]
```diff
@@ -105,3 +105,8 @@ Mock pricer deployed! 🎉
 Transaction hash: 0xebf12fc6603ee857bcabe1c0c6194XX82639a9888776f7ff6e83a409XXXXXXXX
 Deployed contract address: 0x669cC97687c792fc5369d7bdd38cXXXXXXXXXX
 ```
+
+## Verification of Contracts 
+```sh
+truffle run verify ChainlinkPricer@0x17FDF6cE88517bf0Cff19e014E509B543DD78432  --network mainnet
+```
```
Comment (haythem96) [inline scripts/README.md:112]: Why we are adding this to README ?

File: scripts/README.md (hunk 7 above)
Comment (aparnakr) [reply scripts/README.md:29]: yep! This was wrong before! :)

File: scripts/README.md (hunk 8 above)
Comment (aparnakr) [reply scripts/README.md:112]: because I imagine whoever wants to deploy also needs to verify right? Would you add it to packag.json instead? :)

**Pull request closed without merging**